kpatch-build/create-diff-object.c

```diff
@@ -1387,12 +1417,16 @@ static int kpatch_include_hook_elements(struct kpatch_elf *kelf)
 
 	/*
 	 * Strip temporary global load/unload function pointer objects
-	 * used by the kpatch_[load|unload]() macros.
+	 * used by the kpatch_* hook macros.
 	 */
-	list_for_each_entry(sym, &kelf->symbols, list)
-		if (!strcmp(sym->name, "kpatch_load_data") ||
-		    !strcmp(sym->name, "kpatch_unload_data"))
-			sym->include = 0;
+	list_for_each_entry(sym, &kelf->symbols, list) {
+		for (hook_func_ptr = hook_func_ptrs; *hook_func_ptr; hook_func_ptr++) {
+			if (!strcmp(*hook_func_ptr, sym->name)) {
```

Comment:
Despite the comment above this code, they aren't function pointers, they're actually the struct symbol names.  Instead of `hook_func_ptrs`, maybe call them `hook_struct_syms`?

Also, I think there's a bug here, the names are wrong: `kpatch_pre_patch_data` was renamed to `kpatch_pre_patch_data_##_fn` in v2 of the patch.

However, for this code, instead of comparing the struct names, maybe we can just compare `sym->sec->name` with the hook section names, since they will always be in those sections.  That way we don't have to hardcode the struct names at all.

Reply:
Yes, you're right about function pointers vs. struct names.  Also the ##_fn was an artifact of trying to support multiple pre-patch hooks, so I can fix that bug when I clean that up. 

> However, for this code, instead of comparing the struct names, maybe we can just compare sym->sec->name with the hook section names, since they will always be in those sections. That way we don't have to hardcode the struct names at all.

It's been a few weeks since I was in this code, so bear with me :)  The first part of `kpatch_include_hook_elements()` rips through the ELF sections looking for the hook sections (comparing to the `hook_sections[]` strings).  That part would stay as is.

The seond part of `kpatch_include_hook_elements()` strips the hook structures by running through all the ELF symbols and comparing to `hook_func_ptrs[]` ... so you are suggesting that the second set of strings isn't really necessary as we could just look at `sym->sec->name` and compare with the `hook_sections[]` strings?

Reply:
Right :-)